### PR TITLE
Update for Fall 2022 + Add Diffusion Models

### PR DESCRIPTION
**Pull request by anirudh-chakravarthy**

### CODE REVIEW

File: diffusion/inference.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+from unet import Unet
+from utils import save_samples
+
+
+parser = argparse.ArgumentParser(description='Diffusion Model Inference')
+parser.add_argument('--ckpt', required=True, type=str, help="Pretrained checkpoint")
+parser.add_argument('--num-images', default=8, type=int, help="Number of images per iteration")
+parser.add_argument('--sampling-method', default="ddpm", type=str, help="Sampling method (ddpm vs ddim")
+args = parser.parse_args()
+
```
Comment (vanshajc) [inline diffusion/inference.py]: Move this to __main__

File: diffusion/inference.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    """
+    Use the reverse sampling process from DDIM to go from 
+    a tensor of noise to a generated image.
+
+    TODO (Q3.1): fill `ddpm_sample` and the related functions.
+    """
+    num_channels = diffusion_model.channels
+    return diffusion_model.ddpm_sample((num_images, num_channels, image_size, image_size))
+
+def main(
```
Comment (vanshajc) [inline diffusion/inference.py]: all of this can be moved to __main__ block instead of having a main function. let's not specify default paths for the checkpoint either since it won't be part of the repo.

File: diffusion/inference.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+from model import DiffusionModel
+from unet import Unet
+from utils import save_samples
+
+
+parser = argparse.ArgumentParser(description='Diffusion Model Inference')
+parser.add_argument('--ckpt', required=True, type=str, help="Pretrained checkpoint")
+parser.add_argument('--num-images', default=8, type=int, help="Number of images per iteration")
+parser.add_argument('--sampling-method', default="ddpm", type=str, help="Sampling method (ddpm vs ddim")
```
Comment (vanshajc) [inline diffusion/inference.py]: Change to: parser.add_argument('--sampling-method', choices=['ddpm', 'ddim'])

File: diffusion/inference.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import argparse
+import os
+import torch
+
+from model import DiffusionModel
+from unet import Unet
+from utils import save_samples
+
+
+parser = argparse.ArgumentParser(description='Diffusion Model Inference')
```
Comment (vanshajc) [inline diffusion/inference.py]: we need additional arguments for `ddim_sampling_eta` and `ddim_samples`

File: diffusion/inference.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+@torch.no_grad()
+def ddpm_sample(diffusion_model, num_images=8, image_size=32):
+    """
+    Use the reverse sampling process from DDIM to go from 
+    a tensor of noise to a generated image.
+
+    TODO (Q3.1): fill `ddpm_sample` and the related functions.
+    """
+    num_channels = diffusion_model.channels
+    return diffusion_model.ddpm_sample((num_images, num_channels, image_size, image_size))
```
Comment (vanshajc) [inline diffusion/inference.py]: this can be removed I think. it seems the same as the one inside DiffusionModel anyway. Lets call those functions directly.

File: diffusion/model.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        # calculations for posterior q(x_{t-1} | x_t, x_0)
+        self.posterior_variance = self.betas * (
+            1. - self.alphas_cumprod_prev) / (1. - self.alphas_cumprod)
+
+        self.posterior_log_variance_clipped = torch.log(
+            self.posterior_variance.clamp(min =1e-20))
+        self.posterior_mean_coef1 = self.betas * torch.sqrt(
+            self.alphas_cumprod_prev) / (1. - self.alphas_cumprod)
+        self.posterior_mean_coef2 = (1. - self.alphas_cumprod_prev) * torch.sqrt(alphas) / (
+            1. - self.alphas_cumprod)
```
Comment (vanshajc) [inline diffusion/model.py]: we need to blank these, these are all part of the equations

I would try to keep some of the variable names, and add helper functions which we can ask them to fill out.

The only ones I would keep here are:
`alphas, alphas_cumprod, posterior_variance, posterior_mean_coef1 (comment that this is first part of mu in the README), posterior_mean_coef2 (comment that this is the second part of mu in the README)`

For example:

`self.posterior_variance = self._get_posterior_variance(...)`
`self.posterior_log_variance_clipped = torch.log(self.posterior_variance...)`

The rest of the fields, they can create and name however they want. Only those 3 are dependent on the clamping / normalization things.

File: diffusion/unet.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
+from collections import namedtuple
```
Comment (vanshajc) [inline diffusion/unet.py:2]: Add reference at the top

File: diffusion/model.py [hunk 8]
```diff
@@ -31,32 +31,35 @@ def __init__(
 
         alphas = 1. - self.betas
         self.alphas_cumprod = torch.cumprod(alphas, dim=0)
```
Comment (vanshajc) [inline diffusion/model.py]: These should be cleared as well

File: diffusion/model.py [hunk 9]
```diff
@@ -31,32 +31,35 @@ def __init__(
...
         self.sampling_timesteps = default(sampling_timesteps, timesteps) # default num sampling timesteps to number of timesteps at training
 
         assert self.sampling_timesteps <= timesteps
         self.is_ddim_sampling = self.sampling_timesteps < timesteps
         self.ddim_sampling_eta = ddim_sampling_eta
+    
+    def _get_posterior_variance(self):
+        # TODO (Q3.1): compute the variance of the posterior distribution
+        # Hint: this is the \sigma_{t} in the DDPM section
+        pass
```
Comment (vanshajc) [inline diffusion/model.py]: raise NotImplementedError()

**Review by vanshajc — CHANGES_REQUESTED** [review]

**Review by vanshajc — APPROVED** [review]
